### PR TITLE
refactor(storage/reads): refactor and add unit tests for *WindowCountArrayCursor

### PR DESCRIPTION
**Pull request by wolffcm**
This is the first part of a refactoring that makes merges code `integer<type>WindowCountArrayCursor` with `integer<type>CountArrayCursor`. The intent here is that we only want to have to implement aggregate functions in cursors one time (e.g., first, last, max, min) rather than have to implement them once for windowed aggregation and once for aggregation over a whole range. This will let us more easily push down new aggregate functions moving forward.

I also refactored the code in `integer<type>WindowCountArrayCursor` to send out arrays no larger than `MaxPointsPerBlock` (1000) so that we get bounded memory usage and better streaming properties.

This is all part of the issue influxdata/flux#2855. This just handles `count`.  I will do `sum` in a separate PR.

### CODE REVIEW

File: storage/reads/array_cursor.gen.go [hunk 1]
```diff
@@ -200,90 +201,122 @@ func (c floatArraySumCursor) Next() *cursors.FloatArray {
...
 
-	res := cursors.NewIntegerArrayLen(0)
 	rowIdx := 0
 	var acc int64 = 0
 
-	firstTimestamp := a.Timestamps[rowIdx]
-	windowStart := firstTimestamp - firstTimestamp%c.every
-	windowEnd := windowStart + c.every
+	var windowEnd int64
+	if c.every != 0 {
```
Comment (yzhang1991) [inline storage/reads/array_cursor.gen.go:253]: I think you can just use every = INT_MAX in `newIntegerFloatCountArrayCursor`, and eliminate the if here.
If every == INT_MAX, windowStart = 0, windowEnd = 0 + INT_MAX = INT_MAX

File: storage/reads/array_cursor.gen.go [hunk 2]
```diff
@@ -200,90 +201,122 @@ func (c floatArraySumCursor) Next() *cursors.FloatArray {
...
+		windowEnd = math.MaxInt64
+	}
 
 	// enumerate windows
 WINDOWS:
 	for {
 		for ; rowIdx < a.Len(); rowIdx++ {
 			ts := a.Timestamps[rowIdx]
-			if ts >= windowEnd {
+			if c.every != 0 && ts >= windowEnd {
```
Comment (yzhang1991) [inline storage/reads/array_cursor.gen.go:266]: Same as above

File: storage/reads/array_cursor.go [hunk 3]
```diff
@@ -62,47 +63,37 @@ func newSumArrayCursor(cur cursors.Cursor) cursors.Cursor {
...
 	case cursors.BooleanArrayCursor:
-		return &integerBooleanCountArrayCursor{BooleanArrayCursor: cur}
+		return newIntegerBooleanCountArrayCursor(cur)
 	default:
 		panic(fmt.Sprintf("unreachable: %T", cur))
 	}
 }
 
 func newWindowCountArrayCursor(cur cursors.Cursor, req *datatypes.ReadWindowAggregateRequest) cursors.Cursor {
+	if req.WindowEvery == math.MaxInt64 {
```
Comment (yzhang1991) [inline storage/reads/array_cursor.go:81]: This can be eliminated too.

File: storage/reads/array_cursor.gen.go [hunk 4]
```diff
@@ -200,90 +201,122 @@ func (c floatArraySumCursor) Next() *cursors.FloatArray {
 	}
 }
 
-type integerFloatCountArrayCursor struct {
+type integerFloatWindowCountArrayCursor struct {
```
[the remark this answers is not in the document]
Comment (wolffcm) [reply storage/reads/array_cursor.gen.go:204]: I would be okay with that.

File: storage/reads/array_cursor.gen.go [hunk 5]
```diff
@@ -200,90 +201,122 @@ func (c floatArraySumCursor) Next() *cursors.FloatArray {
...
-	return c.FloatArrayCursor.Stats()
-}
-
-func (c *integerFloatCountArrayCursor) Next() *cursors.IntegerArray {
-	a := c.FloatArrayCursor.Next()
-	if len(a.Timestamps) == 0 {
-		return &cursors.IntegerArray{}
+func newIntegerFloatWindowCountArrayCursor(cur cursors.FloatArrayCursor, every int64) *integerFloatWindowCountArrayCursor {
+	resLen := MaxPointsPerBlock
+	if every == 0 {
```
[the remark this answers is not in the document]
Comment (wolffcm) [reply storage/reads/array_cursor.gen.go:213]: I went back and forth on this. I should have explained my reasoning more in the PR comment.

I've a few reasons for doing it this way. First, it is idiomatic to make the zero value useful in Go. Otherwise `0` would just be an invalid value that would cause the code to try to modulo by zero.

Second, using `math.MaxInt64` means that we would get signed integer overflow at this line:
https://github.com/influxdata/influxdb/pull/18354/files#diff-9e2907d1650bef57e89adb1e34e58f5fR263
I'm not sure what the semantics of this are. (The [Go spec](https://golang.org/ref/spec#Integer_overflow) seems a little vague on it.) I'm also unclear on why this ever worked, since if the addition of `math.MaxInt64` was wrapping and producing a negative number, then I would expect this check to always succeed:
https://github.com/influxdata/influxdb/pull/18354/files#diff-9e2907d1650bef57e89adb1e34e58f5fL266
But clearly this was working before, so I'm not sure what was going on.

Given this, I would rather be explicit about when we are aggregating over the whole series and have a little extra logic in there, than be relying on what could be undefined behavior.

Curious to know what you think on this.

File: storage/reads/array_cursor.gen.go (hunk 1 above)
Comment (wolffcm) [reply storage/reads/array_cursor.gen.go:253]: How do you figure that `windowEnd` would be `INT_MAX`? `windowStart` is not likely to be zero.

File: storage/reads/array_cursor.gen.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (yzhang1991) [reply storage/reads/array_cursor.gen.go:213]: > First, it is idiomatic to make the zero value useful in Go. Otherwise `0` would just be an invalid value that would cause the code to try to modulo by zero.

The planner rule guards against the zero value. Normal code paths should never be able to have the zero value case.

> Second, using `math.MaxInt64` means that we would get signed integer overflow at this line:
> https://github.com/influxdata/influxdb/pull/18354/files#diff-9e2907d1650bef57e89adb1e34e58f5fR263

If `every == math.MaxInt64`, for any timestamp `t` that satisfies ` t < math.MaxInt64`:
```
t % every == t
windowStart = t - t % every = t - t = 0
windowEnd = windowStart + every = 0 + math.MaxInt64 = math.MaxInt64
```
If `t == math.MaxInt64`:
```
t % every = math.MaxInt64 % math.MaxInt64 = 0
windowStart = t - t % every = math.MaxInt64 - 0 = math.MaxInt64
windowEnd = windowStart + every = math.MaxInt64 + math.MaxInt64 = -2 // overflow
```
However, `math.MaxInt64` converts to timestamp `2262-04-11 23:47:16.854775807 +0000 UTC`. The closest we can get in Flux is `2262-04-11T23:47:16Z`. We should not be able to get a data point whose timestamp equals `math.MaxInt64` in reality.

Therefore, for any valid timestamp `t`, when `every == math.MaxInt64`, we always have `windowStart == 0, windowEnd == math.MaxInt64`.

[the remark this answers is not in the document]
Comment (yzhang1991) [reply storage/reads/array_cursor.gen.go:213]: I guess we cannot specify a timestamp in line protocol like `2262-04-11 23:47:16.854775807`, can we?

[the remark this answers is not in the document]
Comment (yzhang1991) [reply storage/reads/array_cursor.gen.go:213]: @wolffcm tried and was able to get
```
> t = 2262-04-11T23:47:16.854775807Z
> t
2262-04-11T23:47:16.854775807Z
```
So I tried:
```
import "csv"
inData = "
#datatype,string,long,dateTime:RFC3339,string,string,string,string,double
#group,false,false,false,true,true,true,true,false
#default,_result,,,,,,,
,result,table,_time,_measurement,_field,device,host,_value
,,0,2262-04-11T23:47:16.854775807Z,cpu,percentage,core1,host.local,89.7
"
csv.from(csv: inData)
|> range(start: 2018-05-22T00:00:00Z)
|> filter(fn: (r) => r._measurement == "disk" or r._measurement == "cpu")
|> filter(fn: (r) => r.host == "host.local")
|> window(every: 30s)
|> count()
```

Since the range upper bound is exclusive, you will not be able to get the data point with `math.MaxInt64` timestamp out. And if you don't use `range()`, you are not able to call `window()`.

[the remark this answers is not in the document]
Comment (yzhang1991) [reply storage/reads/array_cursor.gen.go:213]: But for group aggregate a `range()` is not required, then you can run into that `math.MaxInt64` timestamp edge case and miscalculate the window boundaries.
Now I am OK with Chris' current approach and I think we need to have a test case for `math.MaxInt64` timestamp data

**Comment by yzhang1991** [comment]
We discussed offline that the current implementation is good for covering the group aggregate edge case. We just need to add a test case to cover that then this is good.
